---
output: 
  html_document:
    code_download: false
    toc: true                  # table of content true
    toc_depth: 3               # upto three depths of headings (specified by #, ## and ###)
    toc_float: true
    number_sections: true      # if you want number sections at each table header
    theme: united              # many options for theme, this one is my favorite.
    highlight: tango           # specifies the syntax highlighting style
    css: 'style.css'
params:
  phewas_manhattan:  "phewas_man.png"
  phewas_results: "top_results.csv"
  coloc_results: 'None'
  coloc_heatmap: 'None'
title: "`r paste0('Genomics England pheWAS Report' , '') `"
author: ""
date: ""
---


<img src="logo.png" width=100 style="position:absolute;top:4px;bottom:4px;right:4px;" />


```{r include = FALSE}
knitr::opts_chunk$set(echo=FALSE)
```

```{r echo = FALSE}
source("DTable.R")
source("sanitise.R")

suppressPackageStartupMessages({

library(magrittr)
library(DT)
library(data.table)
  
library(htmltools)
library(tidyverse)
library(kableExtra)
   })
```

# PheWAS

## Visualisations

Plots to explore pheWAS results

```{r}
figure_number  <- "Figure 1: "
figure_caption <- "Manhattan plot"
knitr::include_graphics(params$phewas_manhattan, )
htmltools::tags$figcaption( style = 'caption-side: bottom; text-align: center; font-size: 85%%; color: #71879d',
                            htmltools::em(figure_number),
                            htmltools::em(figure_caption))
```


## pheWAS output

```{r}
table   <- as.data.frame(data.table::fread(params$phewas_results))
colnames(table) <- snakecase::to_snake_case(colnames(table))
DTable(table)
```

# Post Analysis

## Colocalization

The following analysis performs genetic colocalization analysis of two potentially related phenotypes, to ask whether they share common genetic causal variant(s). The hypotheses tested are the following:

- H0: neither trait has a genetic association in the region
- H1: only PheWAS phenotype has a genetic association in the region
- H2: only GWAS trait has a genetic association in the region
- H3: both traits are associated, but with different causal variants
- H4: both traits are associated and share a single causal variant

The plot below will display the Posterior Probability (PP) associated to all different hypotheses.

__Below you will find results regarding colocalisation analysis with the GWAS study selected. If no `"coloc"` option was set as `--post_analysis` parameter, then you won't find any further results below.__

### Colocalization output

```{r}
if (params$coloc_results != 'None') {
table   <- as.data.frame(data.table::fread(params$coloc_results))
colnames(table) <- snakecase::to_snake_case(colnames(table))
DTable(table)
}
if (str_detect(params$coloc_results, '.csv') == FALSE){
print('No colocalization analysis was performed. No table displayed')
}
```

### Colocalization heatmap

```{r, results = "asis"}
if (str_detect(params$coloc_heatmap, '.png')) {

  figure_number  <- "Figure 2: "
  figure_caption <- "Posterior Probability (PP) plot"
  cat(paste0("![](", params$coloc_heatmap, ")"), "\n")
  htmltools::tags$figcaption( style = 'caption-side: bottom; text-align: center; font-size: 85%%; color: #71879d',
                              htmltools::em(figure_number),
                              htmltools::em(figure_caption))
}
if (str_detect(params$coloc_heatmap, '.png') == FALSE) {
print('No colocalization analysis was performed. No plot displayed')
}
```
